tests/migrations/test_operations.py

```diff
@@ -5182,6 +5182,319 @@ def assertModelsAndTables(after_db):
+        """
+        Tests removing various types of indexes that include a foreign key.
+        On MySQL such requests might fail due to the index optimizations
+        performed by the DB engine.
+        """
+
+        app_label = "test_cicfkid"
+
+        create_author = migrations.CreateModel(
+            "Author",
+            [("id", models.BigAutoField(primary_key=True))],
+            options={},
+        )
+        author_field = models.ForeignKey(to="Author", on_delete=models.CASCADE)
+        create_book = migrations.CreateModel(
+            "Book",
+            [
+                ("id", models.BigAutoField(primary_key=True)),
+                ("title", models.CharField(max_length=100)),
+            ],
+            options={},
+        )
+
+        project_state = self.apply_operations(
+            app_label,
+            ProjectState(),
+            [
+                create_author,
+                create_book,
+                # add separate operation to test that migration code works with
+                # the FK field that was added after the model creation
+                migrations.AddField("Book", "author", author_field),
+            ],
+        )
+
+        def get_author_field_indexes():
+            Book = project_state.apps.get_model(app_label, "Book")
+            columns = ["author_id"]
+            with connection.schema_editor() as editor:
+                return editor._constraint_names(Book, index=True, column_names=columns)
+
+        def assert_index_created(operations):
+            initial_indexes = get_author_field_indexes()
+            self.apply_operations(app_label, project_state, operations)
+            actual_indexes = get_author_field_indexes()
+
+            # only MySQL implicitly optimizes FK indexes
+            fk_index_optimization_expected = connection.vendor == "mysql"
+            if fk_index_optimization_expected:
+                self.assertNotEqual(set(actual_indexes), set(initial_indexes))
+                # after the additional index found we need to clear field's indexes
+                # before the next checks
+                self.apply_operations(
+                    app_label,
+                    project_state,
+                    [
+                        migrations.RemoveField("Book", "author"),
+                        migrations.AddField("Book", "author", author_field),
+                    ],
+                )
+            else:
+                self.assertEqual(set(actual_indexes), set(initial_indexes))
+
+        assert_index_created(
+            [
+                migrations.AlterUniqueTogether("Book", ["author", "title"]),
+                migrations.AlterUniqueTogether("Book", []),
+            ]
+        )
+        assert_index_created(
+            [
+                migrations.AlterIndexTogether("Book", ["author", "title"]),
+                migrations.AlterIndexTogether("Book", []),
+            ]
+        )
```

Comment:
Twice the same. Did you mean
```suggestion
        assert_index_created(
            [
                migrations.AlterUniqueTogether("Book", ["author", "title"]),
                migrations.AlterUniqueTogether("Book", []),
            ]
        )
        assert_index_created(
            [
                migrations.AlterIndexTogether("Book", []),
                migrations.AlterIndexTogether("Book", ["author", "title"]),
            ]
        )
```
?

Reply:
there are two different operations: AlterUniqueTogether and AlterIndexTogether.
For both we test creating composed index with a FK and then dropping it, to check that both operations behave without failure